viewing_party/party.py

```diff
@@ -1,23 +1,142 @@
+    return avg_rating
+
+
+def get_most_watched_genre(user_data):
+    most_watched_genre = None
+    genre_count_dict = {}
+    if user_data["watched"]:
+        for movie in user_data["watched"]:
+            if movie["genre"] in genre_count_dict:
+                genre_count_dict[movie["genre"]] += 1
+            else:
+                genre_count_dict[movie["genre"]] = 1
+        counts_list = list(genre_count_dict.values())
+        for genre, count in genre_count_dict.items():
+            if count == max(counts_list):
+                most_watched_genre = genre
+
+    return most_watched_genre
+
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
-# -----------------------------------------
-# ------------- WAVE 4 --------------------
-# -----------------------------------------
 
-# -----------------------------------------
-# ------------- WAVE 5 --------------------
-# -----------------------------------------
+def get_unique_watched(user_data):
+    unique_watched = []
+    friends_watched = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_watched.append(movie)
+
+    for movie in user_data["watched"]:
+        if movie not in friends_watched:
+            unique_watched.append(movie)
+
+    return unique_watched
+
+
+def get_friends_unique_watched(user_data):
+    friends_unique_watched = []
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie not in user_data["watched"] and movie not in friends_unique_watched:
+                friends_unique_watched.append(movie)
+
+    return friends_unique_watched
+
+    # -----------------------------------------
+    # ------------- WAVE 4 --------------------
+    # -----------------------------------------
+
+
+def get_available_recs(user_data):
+    recommended = []
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie not in user_data["watched"] and movie["host"] in user_data["subscriptions"]:
+                recommended.append(movie)
+
+    return recommended
+
+    # -----------------------------------------
+    # ------------- WAVE 5 --------------------
+    # -----------------------------------------
+
+
+def get_new_rec_by_genre(user_data):
+    recommended = []
+
+    fav_genre = get_most_watched_genre(user_data)
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie not in user_data["watched"] and movie["genre"] == fav_genre:
+                recommended.append(movie)
+
+    return recommended
+
+
+def get_rec_from_favorites(user_data):
+    recommended = []
+    friends_watched = []
+    if "favorites" in user_data.keys():
+        for friend in user_data["friends"]:
+            for movie in friend["watched"]:
+                friends_watched.append(movie)

```

Comment:
This nested loop to "flatten" out the friends' movies data into a single array was needed quite a few times in this project and would be a great helper function to make!